File: packages/cli/src/commands/validate.ts
/**
 * `tensorcad validate <file|preset>` — the design-rule check.
 *
 * Exits 1 when anything is an error, which makes it usable in CI.
 */

import { formatCount, validate, type Finding, type Severity } from "@tensorcad/core";
import { bool, type Args } from "../args.js";
import { loadDesign } from "../load.js";
import { analysisOptions, impliedGpus } from "../options.js";
import { bold, dim, green, heading, red, writeOut, yellow, blue } from "../format.js";

const ORDER: Severity[] = ["error", "warning", "info"];

const PAINT: Record<Severity, (s: string) => string> = {
  error: red,
  warning: yellow,
  info: blue,
};

function renderGroup(severity: Severity, findings: Finding[]): string[] {
  if (findings.length === 0) return [];
  const paint = PAINT[severity];
  const out: string[] = [heading(paint(`${severity}${findings.length === 1 ? "" : "s"} (${findings.length})`))];
  for (const f of findings) {
    const where = f.path ? `${f.path}${f.port ? `:${f.port}` : ""}` : "";
    out.push(`  ${paint("*")} ${f.message}`);
    out.push(`    ${dim(`${f.rule}${where ? `  ${where}` : ""}`)}`);
    if (f.hint) out.push(`    ${dim("fix:")} ${f.hint}`);
  }
  return out;
}

export function cmdValidate(args: Args): number {
  const { doc } = loadDesign(args._[0]);
  const options = impliedGpus(analysisOptions(args));
  const report = validate(doc, options);

  if (bool(args, "json")) {
    writeOut(
      JSON.stringify(
        {
          name: report.name,
          ok: report.ok,
          counts: report.counts,
          findings: report.findings,
          params: report.analysis.params.total,
        },
        null,
        2,
      ),
    );
    return report.ok ? 0 : 1;
  }

  const out: string[] = [];
  out.push(
    `${bold(report.name)}  ${dim(`${formatCount(report.analysis.params.total)} parameters, T=${report.analysis.options.T}`)}`,
  );

  for (const severity of ORDER) {
    out.push(...renderGroup(severity, report.findings.filter((f) => f.severity === severity)));
  }

  const { error, warning, info } = report.counts;
  out.push("");
  if (report.findings.length === 0) {
    out.push(green("No findings. The design passes every rule."));
  } else {
    const parts = [
      error > 0 ? red(`${error} error${error === 1 ? "" : "s"}`) : null,
      warning > 0 ? yellow(`${warning} warning${warning === 1 ? "" : "s"}`) : null,
      info > 0 ? blue(`${info} info`) : null,
    ].filter(Boolean);
    out.push(`${report.ok ? green("ok") : red("failed")}  ${parts.join(", ")}`);
  }

  writeOut(out.join("\n"));
  return report.ok ? 0 : 1;
}

File: packages/core-go/rules/validate.go
package rules

import (
	"sort"

	"github.com/tensorcad/core/analysis"
	"github.com/tensorcad/core/infer"
	"github.com/tensorcad/core/ir"
)

// Report is what a design-rule check produces.
type Report struct {
	Name     string         `json:"name"`
	Findings []Finding      `json:"findings"`
	Counts   map[string]int `json:"counts"`
	// OK is true when nothing blocks building this design.
	OK bool `json:"ok"`
	// Analysis is the work the rules were run against, returned so a caller
	// that wants both numbers and findings pays for it once.
	Analysis *analysis.Result `json:"analysis"`
	// Overridden is what the document's own severities did, so suppression is
	// never silent: a design cannot drop a finding without the report saying so.
	Overridden []Override `json:"overridden"`
}

// Override is one finding whose severity the document changed.
type Override struct {
	Rule string `json:"rule"`
	Path string `json:"path,omitempty"`
	// From is the severity the rule produced, To what the document asked for.
	// "off" means the finding was dropped.
	From string `json:"from"`
	To   string `json:"to"`
}

var severityOrder = map[string]int{"error": 0, "warning": 1, "info": 2}

// Validate runs every rule over a design and returns the findings sorted by
// severity.
func Validate(doc *ir.Doc, options analysis.Options) (*Report, error) {
	symbols := ir.ResolveSymbols(doc)
	shapeInfo := infer.Shapes(doc, symbols, infer.Options{})
	flat := analysis.Flatten(doc, symbols)
	result, err := analysis.Analyze(doc, options, analysis.Inputs{
		Symbols: symbols, Infer: shapeInfo, Flat: flat,
	})
	if err != nil {
		return nil, err
	}

	ctx := &Ctx{Doc: doc, Symbols: symbols, Infer: shapeInfo, Flat: flat, Analysis: result}
	findings := []Finding{}
	for _, rule := range Rules {
		findings = append(findings, rule.Run(ctx)...)
	}
	for _, message := range flat.Errors {
		findings = append(findings, Finding{Rule: "graph", Severity: "error", Message: message})
	}

	// What the document has decided these rules mean to it, before sorting:
	// a downgraded finding sorts where its new severity puts it.
	overridden := []Override{}
	if len(doc.Rules) > 0 {
		kept := findings[:0]
		for _, f := range findings {
			want, ok := doc.Rules[f.Rule]
			if !ok || want == f.Severity {
				kept = append(kept, f)
				continue
			}
			if !validSeverity(want) {
				// An unreadable override is not an override. Say so rather than
				// guessing, and leave the finding as the rule produced it.
				kept = append(kept, f)
				overridden = append(overridden, Override{
					Rule: f.Rule, Path: f.Path, From: f.Severity, To: "?" + want})
				continue
			}
			overridden = append(overridden, Override{
				Rule: f.Rule, Path: f.Path, From: f.Severity, To: want})
			if want == "off" {
				continue
			}
			f.Severity = want
			kept = append(kept, f)
		}
		findings = kept
	}

	// By severity, then node, then pin; stable, so findings that tie keep the
	// order the rules produced them in.
	//
	// It goes down to the pin rather than stopping at the node because this
	// engine has no port order to inherit: a block's pins live in a map, so two
	// findings on the same block would otherwise be left to tie and come out in
	// whichever order the map handed them over.
	sort.SliceStable(findings, func(i, j int) bool {
		a, b := &findings[i], &findings[j]
		if s := severityOrder[a.Severity] - severityOrder[b.Severity]; s != 0 {
			return s < 0
		}
		if a.Path != b.Path {
			return a.Path < b.Path
		}
		return a.Port < b.Port
	})

	counts := map[string]int{"error": 0, "warning": 0, "info": 0}
	for _, f := range findings {
		counts[f.Severity]++
	}

	return &Report{
		Name:       doc.Meta.Name,
		Findings:   findings,
		Counts:     counts,
		OK:         counts["error"] == 0,
		Analysis:   result,
		Overridden: overridden,
	}, nil
}

// validSeverity is what a document may ask a rule to be.
func validSeverity(s string) bool {
	switch s {
	case "error", "warning", "info", "off":
		return true
	}
	return false
}
